packages/synthetix-main/contracts/storage/Market.sol

```diff
@@ -406,84 +426,137 @@ library Market {
         // Note: The last distributed balance will be cached within this function's execution.
         int256 targetBalance = totalBalance(self);
-        int256 distributedBalance = self.lastDistributedMarketBalance;
-        int256 outstandingBalance = targetBalance - distributedBalance;
+        int256 outstandingBalance = targetBalance - self.lastDistributedMarketBalance;
+
+        (, bool exhaustedUp) = bumpPoolsOut(self, outstandingBalance, maxIter);
+        (, bool exhaustedDown) = bumpPoolsIn(self, outstandingBalance, maxIter);
+
+        if (!exhaustedDown && !exhaustedUp && self.debtDist.totalShares > 0) {
+            self.debtDist.distributeValue(targetBalance - self.lastDistributedMarketBalance);
+            self.lastDistributedMarketBalance = int128(targetBalance);
+        }
+    }
 
-        // Calculate the target value per share of the distribution if it assimilated the market's outstanding balance.
-        int256 targetValuePerShare = self.debtDist.valuePerShare.toLowPrecisionInt128() +
-            outstandingBalance.divDecimal(int128(self.debtDist.totalShares));
+    function bumpPoolsOut(
+        Data storage self,
+        int maxDistributed,
+        uint maxIter
+    ) internal returns (int actuallyDistributed, bool exhausted) {
+        if (maxDistributed <= 0) {
+            return (0, false);
+        }
 
         // Find pools for which this market's max value per share limit is exceeded.
         // Remove them, and distribute their debt up to the limit that is hit.
         // TODO: Polish these comments.
         // Note: This loop should rarely execute the body. When it does, it only executes once for each pool that exceeds the limit since `distributeValue` is not run for most pools. Thus, market users are not hit with any overhead as a result of this.
-        for (uint i = 0; i < maxIter; i++) {
+        uint iters;
+        for (iters = 0; iters < maxIter; iters++) {
             // Exit if there are no in range pools.
             if (self.inRangePools.size() == 0) {
+                //require(self.debtDist.totalShares == 0, "pools in range, but shares in supply");
                 break;
             }
 
+            int targetValuePerShare = self.debtDist.valuePerShare.toLowPrecisionInt128() +
+                (maxDistributed - actuallyDistributed).divDecimal(int128(self.debtDist.totalShares));
+
+            console.log("target value per share", uint(targetValuePerShare));
+            console.log("shares count", uint(self.debtDist.totalShares));
+            console.log("max distributed", uint(self.debtDist.totalShares));
+
             // Exit if the lowest max value per share does not hit the limit.
-            int lowestMaxValuePerShare = -self.inRangePools.getMax().priority;
-            if (lowestMaxValuePerShare >= targetValuePerShare) {
+            HeapUtil.Node memory lowestLimitPool = self.inRangePools.getMax();
+            if (-lowestLimitPool.priority >= targetValuePerShare) {
```

Comment:
Might help to use a declarative value here, or at least comment that `-lowestLimitPool.priority` is the `valuePerShare` of the pool with the lowest limit.

Reply:
added a comment